pkg/export/export.go

```diff
@@ -416,63 +443,96 @@ const (
 )
 
 // ApplyConfig updates the exporter state to the given configuration.
-// Must be called at least once before Export() can be used.
-func (e *Exporter) ApplyConfig(cfg *config.Config) (err error) {
-	// If project_id, location, or cluster were set through the external_labels in the config file,
-	// these values take precedence. If they are unset, the flag value, which defaults to an
-	// environment-specific value on GCE/GKE, is used.
-	builder := labels.NewBuilder(cfg.GlobalConfig.ExternalLabels)
-
-	if !cfg.GlobalConfig.ExternalLabels.Has(KeyProjectID) {
-		builder.Set(KeyProjectID, e.opts.ProjectID)
-	}
-	if !cfg.GlobalConfig.ExternalLabels.Has(KeyLocation) {
-		builder.Set(KeyLocation, e.opts.Location)
+func (e *Exporter) ApplyConfig(cfg *config.Config, opts *ExporterOpts) (err error) {
+	if e.isNopExporter() {
+		return errors.New("nop exporter")
 	}
-	if !cfg.GlobalConfig.ExternalLabels.Has(KeyCluster) {
-		builder.Set(KeyCluster, e.opts.Cluster)
+
+	e.mtx.Lock()
+	defer e.mtx.Unlock()
+
+	optsChanged := false
```

Comment:
suggestion: Don't check for changed status.

Why not just always overwrite the opts?

Reply:
I don't think we should be recreating the metric client each time. If the metric client is recreated, it has to potentially redo the TCP handshake. With HTTP/2, TCP connections are kept alive for a small amount of time to reduce load when multiple requests are made to the same server in succession. In our case, we might send a CMP call every 50ms at the worst case, which is highly likely to benefit from the persistent TPC connection.

Reply:
Done.